feat(evmengine): defer upgrade scheduling to upgrade keeper

0xHansLee opened the pull request:
This PR adds support for rolling upgrades via the `UpgradeEntrypoint` contract. The core implementation is mostly adapted from the `signal` module in Celestia. Instead of introducing a new module, the logic is integrated into the `evmengine` module, which is responsible for handling EVM-triggered upgrades. Since our upgrades are initiated from within the EVM and do not require validator quorum, we were able to adopt only the necessary parts of the `signal` module and implement them within `evmengine` for simplicity.

**Functionally, this is very similar to how we previously supported rolling upgrades using `Fork`.**

As you may know, in Cosmos SDK, a node must not upgrade its binary before the scheduled upgrade height, if the upgrade has already been registered in the `upgrade` keeper. This constraint prevents validators from upgrading early when the upgrade is pre-scheduled.

To address this, we can defer scheduling upgrade. When the `UpgradeEntrypoint` contract emits an upgrade event, the upgrade information is temporarily stored in the `evmengine`'s state. The actual scheduling of the upgrade in the `upgrade` keeper is deferred until the block reaches the specified upgrade height. At that point, during the `PreBlock` phase, the upgrade is scheduled in `upgrade` keeper so that it can be applied within that same block.

issue: #149

stevemilk commented:
Am i able to sync from block 0 to chain tip using only the latest story binary with this feature?

0xHansLee commented:
Removed to use `upgrade-info.json` when app starts up. Previously, the app just observed the latest upgrade because the file only stores the latest upgrade info. Instead, I changed to apply all store upgrades excepting skip ones to support backward-compatible binary that can replay from genesis.

Note: added to dump this `upgrade-info.json` after process the SoftwareUpgrade event which is emitted from EVM for the operator to use cosmovisor to change the binary at the exact height.

cc. @limengformal @lucas2brh

jdubpark commented:
> Removed to use `upgrade-info.json` when app starts up. Previously, the app just observed the latest upgrade because the file only stores the latest upgrade info. Instead, I changed to apply all store upgrades excepting skip ones to support backward-compatible binary that can replay from genesis.
> 
> Note: added to dump this `upgrade-info.json` after process the SoftwareUpgrade event which is emitted from EVM for the operator to use cosmovisor to change the binary at the exact height.

Does Cosmovisor support multiple entries in `upgrade-info.json` now?

0xHansLee commented:
> Does Cosmovisor support multiple entries in upgrade-info.json?

No, they just support the single entry, the latest one.

0xHansLee commented on client/x/evmengine/keeper/upgrades.go:
```diff
@@ -133,10 +145,159 @@ func (k *Keeper) ProcessCancelUpgrade(ctx context.Context, ev *bindings.UpgradeE
+		return errors.Wrap(err, "invalid plan")
+	}
+
+	if plan.Height < ctx.BlockHeight() {
```
This compares the height of upgrade plan for CL with current CL block height, so it is not relevant to the height of EL block.

0xHansLee commented:
@jdubpark Thanks for the notes. Right. Since we trigger the upgrade from EL, quorum of validators for the upgrade is not needed for Story.

> (But all current version upgrades have been triggered on EL by Story, so I suppose the above purpose of the original signal module doesn't matter much for some time.)

I expect we can make our upgrade process be more decentralized by improving our upgrade entry contract on EL.

ramtinms reviewed (approved):
Looks good to me, just to confirm this doesn't have any impact to state sync of older blocks since genesis since we used different way of upgrading the network.

0xHansLee commented:
> Looks good to me, just to confirm this doesn't have any impact to state sync of older blocks since genesis since we used different way of upgrading the network.

It should. But after finding a sync issue in v1.2.1, I have not tested again with the updated branch. I will test again to make sure this works for syncing blocks from the genesis.
